File: common/src/main/java/com/yuanquan/common/utils/AppUpdateUtils.java
package com.yuanquan.common.utils;

import android.app.Application;

import com.cretin.www.cretinautoupdatelibrary.interfaces.AppUpdateInfoListener;
import com.cretin.www.cretinautoupdatelibrary.model.DownloadInfo;
import com.cretin.www.cretinautoupdatelibrary.model.TypeConfig;
import com.cretin.www.cretinautoupdatelibrary.model.UpdateConfig;
import com.cretin.www.cretinautoupdatelibrary.utils.SSLUtils;
import com.yuanquan.common.BuildConfig;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;

public class AppUpdateUtils {
    /**
     * 全局初始化
     *
     * @param context
     */
    public static void init(Application context, Class customActivityClass) {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.connectTimeout(30_000, TimeUnit.SECONDS).readTimeout(30_000, TimeUnit.SECONDS).writeTimeout(30_000, TimeUnit.SECONDS)
                //如果你需要信任所有的证书，可解决根证书不被信任导致无法下载的问题 start
                .sslSocketFactory(SSLUtils.createSSLSocketFactory()).hostnameVerifier(new SSLUtils.TrustAllHostnameVerifier())
                //如果你需要信任所有的证书，可解决根证书不被信任导致无法下载的问题 end
                .retryOnConnectionFailure(true);
        //更新库配置
        UpdateConfig updateConfig = new UpdateConfig().setDebug(BuildConfig.DEBUG)
                .setDataSourceType(TypeConfig.DATA_SOURCE_TYPE_MODEL)
                .setUiThemeType(TypeConfig.UI_THEME_CUSTOM)
                .setCustomActivityClass(customActivityClass)
                .setShowNotification(false)//配置更新的过程中是否在通知栏显示进度
                .setCustomDownloadConnectionCreator(new OkHttp3Connection.Creator(builder));
        com.cretin.www.cretinautoupdatelibrary.utils.AppUpdateUtils.init(context, updateConfig);
    }

    public static void checkUpdate(String apkUrl, int prodVersionCode, String prodVersionName, int forceUpdateFlag, String updateLog, AppUpdateInfoListener appUpdateInfoListener) {
        DownloadInfo info = new DownloadInfo().setApkUrl(apkUrl).setProdVersionCode(prodVersionCode).setProdVersionName(prodVersionName).setForceUpdateFlag(forceUpdateFlag).setUpdateLog(updateLog);
        com.cretin.www.cretinautoupdatelibrary.utils.AppUpdateUtils.getInstance().addAppUpdateInfoListener(appUpdateInfoListener).checkUpdate(info);
    }

    //清除缓存数据
    public static void clearAllData() {
        com.cretin.www.cretinautoupdatelibrary.utils.AppUpdateUtils.getInstance().clearAllData();
    }
}

File: common/src/main/java/com/yuanquan/common/utils/AppUpdateUtils.kt
package com.yuanquan.common.utils

import android.content.Context
import android.content.Intent
import android.text.TextUtils
import com.yuanquan.common.model.DownloadInfo
import java.util.Arrays

/**
 * 可参考一下升级库
 * https://github.com/tangchao0106/AutoUpdateProject
 * implementation 'com.github.MZCretin:AutoUpdateProject:2.0.5'
 */
object AppUpdateUtils {
    @JvmStatic
    fun checkUpdate(
        context: Context,
        apkUrl: String?,
        prodVersionCode: Int,
        prodVersionName: String?,
        forceUpdateFlag: Int,
        updateLog: String?,
    ): DownloadInfo? {
        val info = DownloadInfo().setApkUrl(apkUrl).setProdVersionCode(prodVersionCode)
            .setProdVersionName(prodVersionName).setForceUpdateFlag(forceUpdateFlag)
            .setUpdateLog(updateLog)
        return this.checkUpdate(context, info)
    }

    @JvmStatic
    fun checkUpdate(context: Context, info: DownloadInfo?): DownloadInfo? {
        if (info == null) {
            return null
        }

        //检查当前版本是否需要更新 如果app当前的版本号大于等于线上最新的版本号 不需要升级版本
        val versionCode = SysUtils.getAppVersionCode(context)
        if (versionCode >= info.prodVersionCode) {
            return info
        }

        //如果用户开启了静默下载 其实是否开启强制更新已经没有意义了
        //检查是否强制更新
        if (info.forceUpdateFlag != 0) {
            //需要强制更新
            if (info.forceUpdateFlag == 1) {
                //hasAffectCodes拥有字段强制更新
                val hasAffectCodes = info.hasAffectCodes
                if (!TextUtils.isEmpty(hasAffectCodes)) {
                    val codes = Arrays.asList(
                        *hasAffectCodes.split("\\|".toRegex()).dropLastWhile { it.isEmpty() }
                            .toTypedArray())
                    if (codes.contains(versionCode.toString() + "")) {
                        //包含这个版本 所以需要强制更新
                    } else {
                        //不包含这个版本 所以此版本需要强制更新
                        info.setForceUpdateFlag(0)
                    }
                }
            } else {
                //所有拥有字段强制更新
            }
        }
        return info
    }

    /**
     * 启动Activity
     *
     * @param context
     * @param info
     */
    @JvmStatic
    fun launchActivity(context: Context, info: DownloadInfo?, cla: Class<*>?) {
        val intent = Intent(context, cla)
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP + Intent.FLAG_ACTIVITY_SINGLE_TOP)
        intent.putExtra("info", info)
        context.startActivity(intent)
    }
}